### PR TITLE
perf/solver: don't zero fill weight gradients

### PR DESCRIPTION
**Pull request by alexandermorozov**
**Please verify this PR! I'm not completely sure that it's correct.**

Weight gradients aren't used before they are overwritten at backpropagation
step, so initialization is redundant.

If #89 is applied, this patch improves performance 30% on
`leaf-examples mnist`.

### CODE REVIEW

**Comment by hobofan** [comment]
I highly doubt that it works well when removing this.

The idea of clearing the gradients at that point is that the training of each minibatch starts out with a clean slate. If we don't do that, the gradients accumulate over multiple minibatches leading to unexpected behaviour.

As an alternative the clearing of the weights could be moved to the end of the function, but that would make some diagnostics like looking at the gradients after each minibatch hard/impossible. Personally I prefer the current way between the two alternatives.

**Comment by alexandermorozov** [comment]
Hmm, I was under impression that weight gradients for `i`-th layer are recomputed in the backpropagation pass from the following data:

- output of `i`-th layer (computed during forward pass),
- output gradients of `i`-th layer (computed in the previous backpropagation step for `i+1`-th layer or provided by solver for the last layer),
- weights of `i`-th layer if layer has weigths (computed in the previous epoch),

So weight gradients for `i-th` layer aren't used in computation, so their contents doesn't matter.

I've also tried to verify it in practice, here are graphs for mlp and conv with and without this patch. It looks like there is no visible change.
mlp
conv

**Comment by MichaelHirn** [comment]
Hey @alexandermorozov, thanks for the PR and the charts :+1: 

Max and I agree, that it makes sense to merge the PR as it provides immediate value. If we should decide later to handle the gradients differently, it would be quite easy to put the one line back in.

I am happy to merge!

**Comment by MichaelHirn** [comment]
@homu r+

**Comment by alexandermorozov** [comment]
@MichaelHirn, thanks! It's great that this PR is useful )

I've rebased this branch, it should merge now.

**Comment by hobofan** [comment]
@homu r+
